web/ext/csrf.py

```diff
@@ -0,0 +1,121 @@
+import hashlib, hmac
+
+from binascii import hexlify
+from os import urandom
+from functools import partial
+
+from web.core.util import lazy
+
+from web.security.csrf import CSRFToken, SignedCSRFToken
+
+
+log = __import__('logging').getLogger(__name__)
+
+
+from web.security.csrf import CSRFError
+
+
+class CSRFHandler(object):
+    __slots__ = ('_ctx')
+    
+    def __init__(self, context):
+        self._ctx = context
+        
+    def _get_user_secret(self):
+        return self._ctx.user_csrf.__str__()
+        
+    def verify(self, token):
+        try:
+            SignedCSRFToken(value=token, secret=self._get_user_secret())
+        except CSRFError:
+            return False
+        return True
+        
+    def generate_token(self):
+        return SignedCSRFToken(secret=self._get_user_secret())
+    
+    # Potentially forward getattr? Does that work?
+    def __str__(self):
+        return self.generate_token().signed.decode('utf-8')
+
+
+class CSRFExtension(object):
+    """CSRF Protection extension.
+    
+    This extension offers a process of CSRF protection for forms within an application. It creates 
+    a `context.csrf` object (see `CSRFHandler`) containing the methods needed to get a new token 
+    or verify an existing one.
+    """
+    provides = {'csrf'}
+    
+    def __init__(self, use_session=True, secret=None, cookie=None):

```

Comment:
I contemplated doing it as you suggest here but wasn't sure how you felt, you have told me before explicit is always better than implicit.